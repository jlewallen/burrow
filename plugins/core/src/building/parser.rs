use crate::library::parser::*;

use super::actions::{
    AddScopeAction, BidirectionalDigAction, BuildAreaAction, DuplicateAction, EditAction,
    EditRawAction, LimboAction, MakeItemAction, ObliterateAction,
};

pub struct EditActionParser {}

impl ParsesActions for EditActionParser {
    fn try_parse_action(&self, i: &str) -> EvaluationResult {
        let (_, action) = alt((
            map(
                preceded(pair(tag("edit raw"), spaces), noun_or_specific),
                |item| -> Box<dyn Action> { Box::new(EditRawAction { item }) },
            ),
            map(
                preceded(pair(tag("edit"), spaces), noun_or_specific),
                |item| -> Box<dyn Action> { Box::new(EditAction { item }) },
            ),
        ))(i)?;

        Ok(Some(action))
    }
}

pub struct MakeItemParser {}

impl ParsesActions for MakeItemParser {
    fn try_parse_action(&self, i: &str) -> EvaluationResult {
        let (_, action) = map(
            preceded(
                tuple((tag("@make"), spaces, tag("item"), spaces)),
                string_literal,
            ),
            |name| MakeItemAction { name: name.into() },
        )(i)?;

        Ok(Some(Box::new(action)))
    }
}

pub struct DuplicateActionParser {}

impl ParsesActions for DuplicateActionParser {
    fn try_parse_action(&self, i: &str) -> EvaluationResult {
        let (_, action) = map(
            preceded(pair(tag("@duplicate"), spaces), noun_or_specific),
            |item| DuplicateAction { item },
        )(i)?;

        Ok(Some(Box::new(action)))
    }
}

pub struct LimboActionParser {}

impl ParsesActions for LimboActionParser {
    fn try_parse_action(&self, i: &str) -> EvaluationResult {
        let (_, action) = map(tag("@limbo"), |_| LimboAction {})(i)?;

        Ok(Some(Box::new(action)))
    }
}

pub struct ObliterateActionParser {}

impl ParsesActions for ObliterateActionParser {
    fn try_parse_action(&self, i: &str) -> EvaluationResult {
        let (_, action) = map(
            preceded(pair(tag("@obliterate"), spaces), noun_or_specific),
            |item| ObliterateAction { item },
        )(i)?;

        Ok(Some(Box::new(action)))
    }
}

pub struct BidirectionalDigActionParser {}

impl ParsesActions for BidirectionalDigActionParser {
    fn try_parse_action(&self, i: &str) -> EvaluationResult {
        let (_, action) = map(
            tuple((
                preceded(pair(tag("@dig"), spaces), string_literal),
                preceded(pair(spaces, pair(tag("to"), spaces)), string_literal),
                preceded(pair(spaces, pair(tag("for"), spaces)), string_literal),
            )),
            |(outgoing, returning, new_area)| BidirectionalDigAction {
                outgoing: outgoing.into(),
                returning: returning.into(),
                new_area: new_area.into(),
            },
        )(i)?;

        Ok(Some(Box::new(action)))
    }
}

pub struct ScopeActionParser {}

impl ParsesActions for ScopeActionParser {
    fn try_parse_action(&self, i: &str) -> EvaluationResult {
        let (_, action) = map(
            preceded(pair(tag("@scope"), spaces), camel_case_word),
            |scope_key| AddScopeAction {
                scope_key: scope_key.to_owned(),
            },
        )(i)?;

        Ok(Some(Box::new(action)))
    }
}

pub struct BuildAreaParser {}

impl ParsesActions for BuildAreaParser {
    fn try_parse_action(&self, i: &str) -> EvaluationResult {
        let (_, action) = map(
            preceded(tuple((tag("@build"), spaces)), string_literal),
            |name| BuildAreaAction { name: name.into() },
        )(i)?;

        Ok(Some(Box::new(action)))
    }
}
