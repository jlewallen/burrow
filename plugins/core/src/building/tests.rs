use super::parser::*;
use super::*;
use crate::building::actions::{SaveEntityJsonAction, SaveQuickEditAction};
use crate::building::model::QuickEdit;
use crate::fashion::model::Wearable;
use crate::library::tests::*;
use crate::{
    {carrying::model::Containing, looking::model::new_area_observation, tools},
    {BuildSurroundings, QuickThing},
};

#[test]
fn it_fails_to_edit_unknown_items() -> Result<()> {
    let mut build = BuildSurroundings::new()?;
    let (session, surroundings) = build
        .ground(vec![QuickThing::Object("Cool Broom")])
        .build()?;

    let action = try_parsing(EditActionParser {}, "edit rake")?;
    let action = action.unwrap();
    let reply = action.perform(session, &surroundings)?;

    let reply: SimpleReply = reply.json_as()?;
    assert_eq!(reply, SimpleReply::NotFound);

    Ok(())
}

#[test]
fn it_fails_to_edit_raw_unknown_items() -> Result<()> {
    let mut build = BuildSurroundings::new()?;
    let (session, surroundings) = build
        .ground(vec![QuickThing::Object("Cool Broom")])
        .build()?;

    let action = try_parsing(EditActionParser {}, "edit raw rake")?;
    let action = action.unwrap();
    let reply = action.perform(session, &surroundings)?;

    let reply: SimpleReply = reply.json_as()?;
    assert_eq!(reply, SimpleReply::NotFound);

    Ok(())
}

#[test]
fn it_edits_items_named() -> Result<()> {
    let mut build = BuildSurroundings::new()?;
    let (session, surroundings) = build
        .ground(vec![QuickThing::Object("Cool Broom")])
        .build()?;

    let action = try_parsing(EditActionParser {}, "edit broom")?;
    let action = action.unwrap();
    let reply = action.perform(session, &surroundings)?;

    insta::assert_json_snapshot!(reply.to_debug_json()?);

    Ok(())
}

#[test]
fn it_edits_raw_items_named() -> Result<()> {
    let mut build = BuildSurroundings::new()?;
    let (session, surroundings) = build
        .ground(vec![QuickThing::Object("Cool Broom")])
        .build()?;

    let action = try_parsing(EditActionParser {}, "edit raw broom")?;
    let action = action.unwrap();
    let reply = action.perform(session, &surroundings)?;

    insta::assert_json_snapshot!(reply.to_debug_json()?);

    Ok(())
}

#[test]
fn it_fails_to_edit_items_by_missing_gid() -> Result<()> {
    let mut build = BuildSurroundings::new()?;
    let (session, surroundings) = build
        .ground(vec![QuickThing::Object("Cool Broom")])
        .build()?;

    let action = try_parsing(EditActionParser {}, "edit #1201")?;
    let action = action.unwrap();
    let reply = action.perform(session, &surroundings)?;

    let reply: SimpleReply = reply.json_as()?;
    assert_eq!(reply, SimpleReply::NotFound);

    Ok(())
}

#[test]
fn it_edits_items_by_gid() -> Result<()> {
    let mut build = BuildSurroundings::new()?;
    let (session, surroundings) = build
        .ground(vec![QuickThing::Object("Cool Broom")])
        .build()?;

    let action = try_parsing(EditActionParser {}, "edit #2")?;
    let action = action.unwrap();
    let reply = action.perform(session, &surroundings)?;

    insta::assert_json_snapshot!(reply.to_debug_json()?);

    Ok(())
}

#[test]
fn it_fails_to_duplicate_unknown_items() -> Result<()> {
    let mut build = BuildSurroundings::new()?;
    let (session, surroundings) = build
        .ground(vec![QuickThing::Object("Cool Broom")])
        .build()?;

    let action = try_parsing(DuplicateActionParser {}, "@duplicate rake")?;
    let action = action.unwrap();
    let reply = action.perform(session, &surroundings)?;

    let reply: SimpleReply = reply.json_as()?;
    assert_eq!(reply, SimpleReply::NotFound);

    Ok(())
}

#[test]
fn it_duplicates_items_named() -> Result<()> {
    let mut build = BuildSurroundings::new()?;
    let (session, surroundings) = build
        .hands(vec![QuickThing::Object("Cool Broom")])
        .build()?;

    let action = try_parsing(DuplicateActionParser {}, "@duplicate broom")?;
    let action = action.unwrap();
    let reply = action.perform(session.clone(), &surroundings)?;
    let (_world, person, _area) = surroundings.unpack();

    let reply: SimpleReply = reply.json_as()?;
    assert_eq!(reply, SimpleReply::Done);
    assert_eq!(person.scope::<Containing>()?.holding.len(), 1);
    assert_eq!(
        tools::quantity(&person.scope::<Containing>()?.holding[0].to_entry()?)?,
        2.0
    );

    Ok(())
}

#[test]
fn it_fails_to_obliterate_unknown_items() -> Result<()> {
    let mut build = BuildSurroundings::new()?;
    let (session, surroundings) = build
        .hands(vec![QuickThing::Object("Cool Broom")])
        .build()?;

    let action = try_parsing(ObliterateActionParser {}, "@obliterate rake")?;
    let action = action.unwrap();
    let reply = action.perform(session, &surroundings)?;

    let reply: SimpleReply = reply.json_as()?;
    assert_eq!(reply, SimpleReply::NotFound);

    Ok(())
}

#[test]
fn it_obliterates_items_named() -> Result<()> {
    let mut build = BuildSurroundings::new()?;
    let (session, surroundings) = build
        .hands(vec![QuickThing::Object("Cool Broom")])
        .build()?;

    let action = try_parsing(ObliterateActionParser {}, "@obliterate broom")?;
    let action = action.unwrap();
    let reply = action.perform(session.clone(), &surroundings)?;
    let (_world, person, area) = surroundings.unpack();

    let reply: SimpleReply = reply.json_as()?;
    assert_eq!(reply, SimpleReply::Done);
    // It's not enough just to check this, but why not given how easy.
    // Should actually verify it's deleted.
    assert_eq!(person.scope::<Containing>()?.holding.len(), 0);
    assert_eq!(area.scope::<Containing>()?.holding.len(), 0);

    build.flush()?;

    Ok(())
}

#[test]
fn it_digs_bidirectionally() -> Result<()> {
    let mut build = BuildSurroundings::new()?;
    let (session, surroundings) = build.plain().build()?;

    let action = try_parsing(
        BidirectionalDigActionParser {},
        r#"@dig "North Exit" to "South Exit" for "New Area""#,
    )?;
    let action = action.unwrap();
    let reply = action.perform(session.clone(), &surroundings)?;
    let (_, living, _area) = surroundings.unpack();

    // Not the best way of finding the constructed area.
    let destination = session
        .entry(&LookupBy::Gid(&EntityGid::new(4)))?
        .ok_or(DomainError::EntityNotFound)?;

    let reply: AreaObservation = reply.json_as()?;
    assert_eq!(reply, new_area_observation(&living, &destination)?);

    Ok(())
}

#[test]
fn it_makes_items() -> Result<()> {
    let mut build = BuildSurroundings::new()?;
    let (session, surroundings) = build.plain().build()?;

    let action = try_parsing(MakeItemParser {}, r#"@make item "Blue Rake""#)?;
    let action = action.unwrap();
    let reply = action.perform(session.clone(), &surroundings)?;
    let (_, living, _area) = surroundings.unpack();

    let reply: SimpleReply = reply.json_as()?;
    assert_eq!(reply, SimpleReply::Done);

    assert_eq!(living.scope::<Containing>()?.holding.len(), 1);

    Ok(())
}

#[test]
fn it_saves_changes_to_description() -> Result<()> {
    let mut build = BuildSurroundings::new()?;
    let (session, surroundings) = build.plain().build()?;
    let (_world, _living, _area) = surroundings.unpack();

    let description = "Would be really weird if this was the original description".to_owned();
    let mut quick_edit = QuickEdit::default();
    quick_edit.name = Some("NAME!".to_owned());
    quick_edit.desc = Some(description.to_owned());

    let action = Box::new(SaveQuickEditAction {
        key: EntityKey::new("world"),
        copy: WorkingCopy::Markdown(quick_edit.to_string()),
    });
    let reply = action.perform(session.clone(), &surroundings)?;
    let (world, _living, _area) = surroundings.unpack();

    let reply: SimpleReply = reply.json_as()?;
    assert_eq!(reply, SimpleReply::Done);

    assert_eq!(world.desc()?.unwrap(), description.as_str());

    Ok(())
}

#[test]
fn it_saves_changes_to_whole_entities() -> Result<()> {
    let mut build = BuildSurroundings::new()?;
    let (session, surroundings) = build.plain().build()?;
    let (_, living, area) = surroundings.unpack();

    let original = living.entity();
    let original = original.to_json_value()?;

    let action = Box::new(SaveEntityJsonAction {
        key: area.key().clone(),
        copy: WorkingCopy::Json(original),
    });
    let reply = action.perform(session.clone(), &surroundings)?;
    let (_, _living, _area) = surroundings.unpack();

    let reply: SimpleReply = reply.json_as()?;
    assert_eq!(reply, SimpleReply::Done);

    // TODO Would be really nice to have some assurances here, even though
    // I'm wondering how often this will actually get used.

    Ok(())
}

#[test]
fn it_adds_scopes_to_solo_held_items() -> Result<()> {
    let mut build = BuildSurroundings::new()?;
    let jacket = build.entity()?.named("Jacket")?.carryable()?.into_entry()?;

    assert!(!jacket.has_scope::<Wearable>()?);

    let (session, surroundings) = build
        .plain()
        .hands(vec![QuickThing::Actual(jacket.clone())])
        .build()?;

    let action = try_parsing(ScopeActionParser {}, r#"@scope wearable"#)?.unwrap();
    let reply = action.perform(session.clone(), &surroundings)?;
    let (_, living, _area) = surroundings.unpack();

    let reply: SimpleReply = reply.json_as()?;
    assert_eq!(reply, SimpleReply::Done);

    assert_eq!(living.scope::<Containing>()?.holding.len(), 1);

    assert!(jacket.has_scope::<Wearable>()?);

    Ok(())
}
