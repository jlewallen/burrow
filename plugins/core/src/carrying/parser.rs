use super::actions::*;
use crate::library::parser::*;

pub struct HoldActionParser {}

impl ParsesActions for HoldActionParser {
    fn try_parse_action(&self, i: &str) -> EvaluationResult {
        let (_, action) = map(
            separated_pair(tag("hold"), spaces, alt((quantified, noun))),
            |(_, item)| HoldAction { item },
        )(i)?;

        Ok(Some(Box::new(action)))
    }
}

pub struct DropActionParser {}

impl ParsesActions for DropActionParser {
    fn try_parse_action(&self, i: &str) -> EvaluationResult {
        let specific = map(
            separated_pair(tag("drop"), spaces, alt((quantified, noun))),
            |(_, item)| DropAction {
                maybe_item: Some(Item::Held(Box::new(item))),
            },
        );

        let everything = map(tag("drop"), |_| DropAction { maybe_item: None });

        let (_, action) = alt((specific, everything))(i)?;

        Ok(Some(Box::new(action)))
    }
}

pub struct TakeOutActionParser {}

impl ParsesActions for TakeOutActionParser {
    fn try_parse_action(&self, i: &str) -> EvaluationResult {
        let item = map(separated_pair(tag("take"), spaces, noun), |(_, item)| item);

        let (_, action) = map(
            separated_pair(separated_pair(item, spaces, tag("out of")), spaces, noun),
            |(item, vessel)| TakeOutAction {
                item: Item::Contained(Box::new(item.0)),
                vessel: vessel,
            },
        )(i)?;

        Ok(Some(Box::new(action)))
    }
}

pub struct PutInsideActionParser {}

impl ParsesActions for PutInsideActionParser {
    fn try_parse_action(&self, i: &str) -> EvaluationResult {
        let item = map(separated_pair(tag("put"), spaces, noun), |(_, item)| item);

        let (_, action) = map(
            separated_pair(
                separated_pair(
                    item,
                    spaces,
                    pair(tag("inside"), opt(pair(spaces, tag("of")))),
                ),
                spaces,
                noun,
            ),
            |(item, vessel)| PutInsideAction {
                item: item.0,
                vessel: vessel,
            },
        )(i)?;

        Ok(Some(Box::new(action)))
    }
}

pub struct GiveToActionParser {}

impl ParsesActions for GiveToActionParser {
    fn try_parse_action(&self, i: &str) -> EvaluationResult {
        let (_, action) = map(
            separated_pair(
                preceded(tag("give"), preceded(spaces, noun)),
                spaces,
                preceded(tag("to"), preceded(spaces, person)),
            ),
            |(item, receiver)| GiveToAction {
                item: Item::Held(Box::new(item)),
                receiver,
            },
        )(i)?;

        Ok(Some(Box::new(action)))
    }
}
