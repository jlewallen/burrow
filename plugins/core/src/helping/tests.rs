use insta::assert_json_snapshot;

use super::parser::*;
use crate::library::tests::*;

#[test]
fn it_reads_default_help() -> Result<()> {
    let (_surroundings, effect) = parse_and_perform(ReadHelpParser {}, "help")?;

    assert_json_snapshot!(effect.to_debug_json()?);

    Ok(())
}

#[test]
fn it_reads_help_by_name() -> Result<()> {
    let (_surroundings, effect) = parse_and_perform(ReadHelpParser {}, "help Food")?;

    assert_json_snapshot!(effect.to_debug_json()?);

    Ok(())
}

#[test]
fn it_allows_editing_default_help() -> Result<()> {
    let (_surroundings, effect) = parse_and_perform(HelpWithParser {}, "edit help")?;

    assert_json_snapshot!(effect.to_debug_json()?);

    Ok(())
}

#[test]
fn it_allows_editing_help_by_name() -> Result<()> {
    let (_surroundings, effect) = parse_and_perform(HelpWithParser {}, "edit help Food")?;

    assert_json_snapshot!(effect.to_debug_json()?);

    Ok(())
}
