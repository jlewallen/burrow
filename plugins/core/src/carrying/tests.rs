use super::parser::*;
use super::*;
use crate::carrying::model::{Carryable, Containing};
use crate::library::tests::*;
use crate::location::Location;

#[test]
fn it_holds_unheld_items() -> Result<()> {
    let mut build = BuildSurroundings::new()?;
    let (session, surroundings) = build
        .ground(vec![QuickThing::Object("Cool Rake")])
        .build()?;

    let (_, person, area) = surroundings.unpack();
    assert_eq!(person.scope::<Containing>()?.unwrap().holding.len(), 0);
    assert_eq!(area.scope::<Containing>()?.unwrap().holding.len(), 1);

    let action = try_parsing(HoldActionParser {}, "hold rake")?;
    let action = action.unwrap();
    let effect = action.perform(session.clone(), &surroundings)?;

    assert_eq!(effect, Effect::Ok);

    assert_eq!(person.scope::<Containing>()?.unwrap().holding.len(), 1);
    assert_eq!(area.scope::<Containing>()?.unwrap().holding.len(), 0);

    build.close()?;

    Ok(())
}

#[test]
fn it_separates_multiple_ground_items_when_held() -> Result<()> {
    let mut build = BuildSurroundings::new()?;
    let (session, surroundings) = build
        .ground(vec![QuickThing::Multiple("Cool Rake", 2.0)])
        .build()?;

    let (_, person, area) = surroundings.unpack();
    assert_eq!(person.scope::<Containing>()?.unwrap().holding.len(), 0);
    assert_eq!(area.scope::<Containing>()?.unwrap().holding.len(), 1);

    let action = try_parsing(HoldActionParser {}, "hold rake")?;
    let action = action.unwrap();
    let effect = action.perform(session.clone(), &surroundings)?;

    assert_eq!(effect, Effect::Ok);

    let held = &person.scope::<Containing>()?.unwrap().holding;
    let ground = &area.scope::<Containing>()?.unwrap().holding;
    assert_eq!(held.len(), 1);
    assert_eq!(ground.len(), 1);

    let held_keys: HashSet<_> = held.iter().map(|i| i.key().clone()).collect();
    let ground_keys: HashSet<_> = ground.iter().map(|i| i.key().clone()).collect();
    let common_keys: HashSet<_> = held_keys.intersection(&ground_keys).collect();
    assert_eq!(common_keys.len(), 0);

    build.close()?;

    Ok(())
}

#[test]
fn it_combines_multiple_items_when_together_on_ground() -> Result<()> {
    let mut build = BuildSurroundings::new()?;
    let same_kind = build.make(QuickThing::Object("Cool Rake"))?;
    tools::set_quantity(&same_kind, &2.0.into())?;
    let (first, second) = tools::separate(&same_kind, &1.0.into())?;
    let (session, surroundings) = build
        .ground(vec![QuickThing::Actual(first.clone())])
        .hands(vec![QuickThing::Actual(second)])
        .build()?;

    let (_, person, area) = surroundings.unpack();
    assert_eq!(person.scope::<Containing>()?.unwrap().holding.len(), 1);
    assert_eq!(area.scope::<Containing>()?.unwrap().holding.len(), 1);

    let action = try_parsing(HoldActionParser {}, "hold rake")?;
    let action = action.unwrap();
    let effect = action.perform(session.clone(), &surroundings)?;

    assert_eq!(effect, Effect::Ok);

    assert_eq!(person.scope::<Containing>()?.unwrap().holding.len(), 1);
    assert_eq!(area.scope::<Containing>()?.unwrap().holding.len(), 0);

    build.close()?;

    Ok(())
}

#[test]
fn it_fails_to_hold_unknown_items() -> Result<()> {
    let mut build = BuildSurroundings::new()?;
    let (session, surroundings) = build
        .ground(vec![QuickThing::Object("Cool Broom")])
        .build()?;

    let action = try_parsing(HoldActionParser {}, "hold rake")?;
    let action = action.unwrap();
    let reply = action.perform(session.clone(), &surroundings)?;
    let (_, person, area) = surroundings.unpack();

    let reply: SimpleReply = reply.json_as()?;
    assert_eq!(reply, SimpleReply::NotFound);

    assert_eq!(person.scope::<Containing>()?.unwrap().holding.len(), 0);
    assert_eq!(area.scope::<Containing>()?.unwrap().holding.len(), 1);

    build.close()?;

    Ok(())
}

#[test]
fn it_drops_held_items() -> Result<()> {
    let mut build = BuildSurroundings::new()?;
    let (session, surroundings) = build.hands(vec![QuickThing::Object("Cool Rake")]).build()?;

    let action = try_parsing(DropActionParser {}, "drop rake")?;
    let action = action.unwrap();
    let effect = action.perform(session.clone(), &surroundings)?;

    assert_eq!(effect, Effect::Ok);

    let (_, person, area) = surroundings.unpack();

    assert_eq!(person.scope::<Containing>()?.unwrap().holding.len(), 0);
    assert_eq!(area.scope::<Containing>()?.unwrap().holding.len(), 1);

    build.close()?;

    Ok(())
}

#[test]
fn it_fails_to_drop_unknown_items() -> Result<()> {
    let mut build = BuildSurroundings::new()?;
    let (session, surroundings) = build
        .hands(vec![QuickThing::Object("Cool Broom")])
        .build()?;

    let action = try_parsing(DropActionParser {}, "drop rake")?;
    let action = action.unwrap();
    let reply = action.perform(session.clone(), &surroundings)?;
    let (_, person, area) = surroundings.unpack();

    let reply: SimpleReply = reply.json_as()?;
    assert_eq!(reply, SimpleReply::NotFound);

    assert_eq!(person.scope::<Containing>()?.unwrap().holding.len(), 1);
    assert_eq!(area.scope::<Containing>()?.unwrap().holding.len(), 0);

    build.close()?;

    Ok(())
}

#[test]
fn it_fails_to_drop_unheld_items() -> Result<()> {
    let mut build = BuildSurroundings::new()?;
    let (session, surroundings) = build
        .ground(vec![QuickThing::Object("Cool Broom")])
        .build()?;

    let action = try_parsing(DropActionParser {}, "drop rake")?;
    let action = action.unwrap();
    let reply = action.perform(session.clone(), &surroundings)?;
    let (_, person, area) = surroundings.unpack();

    let reply: SimpleReply = reply.json_as()?;
    assert_eq!(reply, SimpleReply::NotFound);

    assert_eq!(person.scope::<Containing>()?.unwrap().holding.len(), 0);
    assert_eq!(area.scope::<Containing>()?.unwrap().holding.len(), 1);

    build.close()?;

    Ok(())
}

#[test]
fn it_fails_to_put_item_in_non_containers() -> Result<()> {
    let mut build = BuildSurroundings::new()?;
    let vessel = build
        .entity()?
        .named("Not A Vessel")?
        .save()?
        .carryable()?
        .into_entity()?;
    let (session, surroundings) = build
        .hands(vec![
            QuickThing::Object("key"),
            QuickThing::Actual(vessel.clone()),
        ])
        .build()?;

    let action = try_parsing(PutInsideActionParser {}, "put key inside vessel")?;
    let action = action.unwrap();
    let reply = action.perform(session.clone(), &surroundings)?;
    let (_world, person, _area) = surroundings.unpack();

    insta::assert_json_snapshot!(reply.to_debug_json()?);

    assert_eq!(person.scope::<Containing>()?.unwrap().holding.len(), 2);
    assert!(vessel.scope::<Containing>()?.is_none());

    build.close()?;

    Ok(())
}

#[test]
fn it_puts_items_in_containers() -> Result<()> {
    let mut build = BuildSurroundings::new()?;
    let vessel = build
        .entity()?
        .named("Vessel")?
        .save()?
        .carryable()?
        .holding(&vec![])?
        .into_entity()?;
    let (session, surroundings) = build
        .hands(vec![
            QuickThing::Object("key"),
            QuickThing::Actual(vessel.clone()),
        ])
        .build()?;

    let action = try_parsing(PutInsideActionParser {}, "put key inside vessel")?;
    let action = action.unwrap();
    let reply = action.perform(session.clone(), &surroundings)?;
    let (_world, person, _area) = surroundings.unpack();

    insta::assert_json_snapshot!(reply.to_debug_json()?);

    assert_eq!(person.scope::<Containing>()?.unwrap().holding.len(), 1);
    assert_eq!(vessel.scope::<Containing>()?.unwrap().holding.len(), 1);

    build.close()?;

    Ok(())
}

#[test]
fn it_takes_items_out_of_containers() -> Result<()> {
    let mut build = BuildSurroundings::new()?;
    let key = build
        .entity()?
        .named("Key")?
        .save()?
        .carryable()?
        .into_entity()?;
    let vessel = build
        .entity()?
        .named("Vessel")?
        .save()?
        .carryable()?
        .holding(&vec![key.clone()])?
        .into_entity()?;
    let (session, surroundings) = build
        .hands(vec![QuickThing::Actual(vessel.clone())])
        .build()?;

    let action = try_parsing(TakeOutActionParser {}, "take key out of vessel")?;
    let action = action.unwrap();
    let reply = action.perform(session.clone(), &surroundings)?;
    let (_world, person, _area) = surroundings.unpack();

    insta::assert_json_snapshot!(reply.to_debug_json()?);

    assert_eq!(person.scope::<Containing>()?.unwrap().holding.len(), 2);
    assert_eq!(vessel.scope::<Containing>()?.unwrap().holding.len(), 0);
    assert_eq!(
        *key.scope::<Location>()?
            .unwrap()
            .container
            .as_ref()
            .unwrap()
            .key(),
        person.key()
    );

    build.close()?;

    Ok(())
}

#[test]
fn it_gives_items_to_others() -> Result<()> {
    let mut build = BuildSurroundings::new()?;
    let carla = build.with(build_entity().living().name("Carla").try_into()?)?;
    let (session, surroundings) = build
        .hands(vec![QuickThing::Object("key")])
        .occupying(vec![QuickThing::Actual(carla.clone())])
        .build()?;

    let action = try_parsing(GiveToActionParser {}, "give key to Carla")?;
    let action = action.unwrap();
    let reply = action.perform(session.clone(), &surroundings)?;
    let (_world, person, _area) = surroundings.unpack();

    assert_eq!(person.scope::<Containing>()?.unwrap().holding.len(), 0);
    assert_eq!(carla.scope::<Containing>()?.unwrap().holding.len(), 1);

    insta::assert_json_snapshot!(reply.to_debug_json()?);

    build.close()?;

    Ok(())
}

#[test]
fn it_drops_quantified_items() -> Result<()> {
    let mut build = BuildSurroundings::new()?;
    let (session, surroundings) = build
        .hands(vec![QuickThing::Multiple("Coin", 10.0)])
        .build()?;

    let action = try_parsing(DropActionParser {}, "drop 2 coin")?;
    let action = action.unwrap();
    let effect = action.perform(session.clone(), &surroundings)?;

    assert_eq!(effect, Effect::Ok);

    let (_, person, area) = surroundings.unpack();

    assert_eq!(person.scope::<Containing>()?.unwrap().holding.len(), 1);
    assert_eq!(area.scope::<Containing>()?.unwrap().holding.len(), 1);

    build.close()?;

    Ok(())
}

#[test]
fn it_fails_to_drop_when_available_quantity_insufficient() -> Result<()> {
    let mut build = BuildSurroundings::new()?;
    let (session, surroundings) = build
        .hands(vec![QuickThing::Multiple("Coin", 4.0)])
        .build()?;

    let action = try_parsing(DropActionParser {}, "drop 5 coin")?;
    let action = action.unwrap();
    let effect = action.perform(session.clone(), &surroundings)?;

    let reply: SimpleReply = effect.json_as()?;
    assert_eq!(reply, SimpleReply::NotFound);

    let (_, person, area) = surroundings.unpack();

    assert_eq!(person.scope::<Containing>()?.unwrap().holding.len(), 1);
    assert_eq!(area.scope::<Containing>()?.unwrap().holding.len(), 0);

    build.close()?;

    Ok(())
}

#[test]
fn it_drops_specified_quantity() -> Result<()> {
    let mut build = BuildSurroundings::new()?;
    let (session, surroundings) = build
        .hands(vec![QuickThing::Multiple("Coin", 4.0)])
        .build()?;

    let (_, person, _) = surroundings.unpack();

    assert_eq!(
        person.scope::<Containing>()?.unwrap().holding[0]
            .to_entity()?
            .scope::<Carryable>()?
            .unwrap()
            .quantity(),
        4.0
    );

    let action = try_parsing(DropActionParser {}, "drop 2 coin")?;
    let action = action.unwrap();
    let effect = action.perform(session.clone(), &surroundings)?;

    assert_eq!(effect, Effect::Ok);

    let (_, person, area) = surroundings.unpack();

    assert_eq!(person.scope::<Containing>()?.unwrap().holding.len(), 1);
    assert_eq!(area.scope::<Containing>()?.unwrap().holding.len(), 1);

    assert_eq!(
        person.scope::<Containing>()?.unwrap().holding[0]
            .to_entity()?
            .scope::<Carryable>()?
            .unwrap()
            .quantity(),
        2.0
    );
    assert_eq!(
        area.scope::<Containing>()?.unwrap().holding[0]
            .to_entity()?
            .scope::<Carryable>()?
            .unwrap()
            .quantity(),
        2.0
    );

    build.close()?;

    Ok(())
}

#[test]
fn it_fails_to_hold_when_available_quantity_insufficient() -> Result<()> {
    let mut build = BuildSurroundings::new()?;
    let (session, surroundings) = build
        .ground(vec![QuickThing::Multiple("Coin", 4.0)])
        .build()?;

    let action = try_parsing(HoldActionParser {}, "hold 5 coin")?;
    let action = action.unwrap();
    let effect = action.perform(session.clone(), &surroundings)?;

    let reply: SimpleReply = effect.json_as()?;
    assert_eq!(reply, SimpleReply::NotFound);

    let (_, person, area) = surroundings.unpack();

    assert_eq!(person.scope::<Containing>()?.unwrap().holding.len(), 0);
    assert_eq!(area.scope::<Containing>()?.unwrap().holding.len(), 1);

    build.close()?;

    Ok(())
}

#[test]
fn it_holds_specified_quantity() -> Result<()> {
    let mut build = BuildSurroundings::new()?;
    let (session, surroundings) = build
        .ground(vec![QuickThing::Multiple("Coin", 4.0)])
        .build()?;

    let (_, _, area) = surroundings.unpack();

    assert_eq!(
        area.scope::<Containing>()?.unwrap().holding[0]
            .to_entity()?
            .scope::<Carryable>()?
            .unwrap()
            .quantity(),
        4.0
    );

    let action = try_parsing(HoldActionParser {}, "hold 2 coin")?;
    let action = action.unwrap();
    let effect = action.perform(session.clone(), &surroundings)?;

    assert_eq!(effect, Effect::Ok);

    let (_, person, area) = surroundings.unpack();

    assert_eq!(person.scope::<Containing>()?.unwrap().holding.len(), 1);
    assert_eq!(area.scope::<Containing>()?.unwrap().holding.len(), 1);

    assert_eq!(
        person.scope::<Containing>()?.unwrap().holding[0]
            .to_entity()?
            .scope::<Carryable>()?
            .unwrap()
            .quantity(),
        2.0
    );
    assert_eq!(
        area.scope::<Containing>()?.unwrap().holding[0]
            .to_entity()?
            .scope::<Carryable>()?
            .unwrap()
            .quantity(),
        2.0
    );

    build.close()?;

    Ok(())
}

#[test]
#[ignore]
fn it_trades_one_for_one() -> Result<()> {
    let mut build = BuildSurroundings::new()?;
    let rake = build.with(build_entity().name("Rake"))?;
    let carla = build
        .entity()?
        .named("Carla")?
        .save()?
        .holding(&vec![rake])?
        .into_entity()?;
    let (session, surroundings) = build
        .hands(vec![QuickThing::Object("key")])
        .occupying(vec![QuickThing::Actual(carla.clone())])
        .build()?;

    let action = try_parsing(TradeActionParser {}, "trade Carla key for rake")?;
    let action = action.unwrap();
    let effect = action.perform(session.clone(), &surroundings)?;

    assert_eq!(effect, Effect::Ok);

    let (_, person, _) = surroundings.unpack();

    assert_eq!(person.scope::<Containing>()?.unwrap().holding.len(), 1);
    assert_eq!(carla.scope::<Containing>()?.unwrap().holding.len(), 1);

    build.close()?;

    Ok(())
}
