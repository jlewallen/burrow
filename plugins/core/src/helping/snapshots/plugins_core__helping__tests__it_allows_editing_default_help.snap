---
source: plugins/core/src/helping/tests.rs
expression: effect.to_debug_json()?
---
{
  "reply": {
    "taggedJson": {
      "editorReply": {
        "editing": {
          "markdown": "# Hello, world!"
        },
        "key": "E-0",
        "save": {
          "saveHelpAction": {
            "copy": {
              "markdown": "!#$value"
            },
            "key": "E-0"
          }
        }
      }
    }
  }
}
